app/controllers/orders_controller.rb

```diff
@@ -1,16 +1,40 @@
 class OrdersController < ApplicationController
+  before_action :find_cart_order
+
   def view_cart
-    @order = Order.find_by(id: session[:cart_id])
   end
 
   def checkout
-    @order = Order.find_by(id: session[:cart_id])
     unless @order
       flash[:error] = "Unable to checkout cart"
       redirect_to cart_path
     end
   end
 
   def purchase
+    if @order && @order.update(order_params) && @order.cart_errors.empty?
+      @order.cart_checkout
+      @order.update(status: "complete", cc_four: params[:order][:cc_all][-4..-1]) # front end valid. on form for min 4 chars
```

Comment:
Does this verification logic belong here? Since it's kind of a validation, and kind of business logic, I'm wondering if it should go in the model. What do you think? Also, nice use of [-4..-1]. Very clean. :-)